Content modelling/417 Get embedded content by target_content_id

tahb opened the pull request:
# What does this do?

Given a `target_content_id` which is the edition of the content block/embeddable content, it will find and return any other live editions [where it has been embedded and a link set up](https://github.com/alphagov/publishing-api/pull/2813).

For The content Object store within Whitehall we only need a few fields compared to the existing `content/:id` endpoint: `title`, `base_path`, `document_type`, `target_content_id` and a couple of fields about the associated organisation.

With this data we’ll present a table that allows reusable content publishers to see where their content is currently being included.

# Why a new endpoint?

We have previously tried to use [the `/v2/content/:content_id` endpoint](https://github.com/alphagov/publishing-api/blob/main/docs/api.md#put-v2contentcontent_id). Though it includes a lot more `Edition` data than we need it also returned the publishing organisation link and gave us pagination and sorting for free. **This worked locally but on[ Integration this timed out](https://gds.slack.com/archives/C0776B04EJU/p1723559847781489) and our implementation (which is flagged off in production) is currently broken.**

We tracked down the performance problem to [a subquery used](https://github.com/alphagov/publishing-api/blob/3f4cc8523b1713aefc1e8f70950e1a81cfa56f22/app/presenters/queries/content_item_presenter.rb#L190) when the `link_set_links` `field` is passed through. [It appears we added this a few weeks ago in an effort to extend the endpoint for our purposes](https://github.com/alphagov/publishing-api/pull/2823). As a result of the data that follows I think this change should be reverted and [I'll propose it as a separate change](https://github.com/alphagov/publishing-api/pull/2847).

We did some local [benchmarking](https://ruby-doc.org/stdlib-2.7.1/libdoc/benchmark/rdoc/Benchmark.html) and found the query was slowing down notably after a few hundred thousand records:

```
2024-08-15 17:24:20 Document count: 178924
2024-08-15 17:24:20 Edition count: 178925
2024-08-15 17:24:20 Link count: 178855
                                      user            system      total          real
2024-08-15 17:24:20   0.003363   0.012408   0.015771 (  0.334085)
1.92s in the browser

2024-08-15 17:29:01 Document count: 228173
2024-08-15 17:29:01 Edition count: 228174
2024-08-15 17:29:01 Link count: 228106
                                      user            system      total          real
2024-08-15 17:29:01   0.019349   0.013283   0.032632 (  0.499379)
3.8s in the browser
```

We stopped seeding at this point as we could see that seeding more would eventually result in a timeout.

On integration the comparable data counts at the time were:

```
39_953_449 links
14_170_691 editions
1_323_401 documents 
931_902 link sets
```

It seems reasonable that there might be more than hundreds of thousands of
documents and millions of links in production. We can't find an obvious
fault in the existing subquery. Instead of risking a change to it we
create a new endpoint which does just the thing we need.

On the same data set we see a marked improvement:
```
2024-08-16 11:19:27 Document: 230366
2024-08-16 11:19:27 Edition: 230365
2024-08-16 11:19:27 Link: 230295
                                      user            system      total          real
2024-08-16 11:19:27   0.000593   0.000170   0.000763 (  0.001478)
```

A negative side effect of this is that we won't get pagination and
sorting for free.

## Naming

We continue to use the relatively new term of `embedded`. [It was the language used on the original linking work](https://github.com/alphagov/publishing-api/pull/2813) and also is what we use as our embed references: `{{embed…}}`.

We sit alongside many other endpoints and services which are named "link", "linkables", "link sets" and "link set links". Though we of course use `links` under the hood to find embedded content we try to abstract that in order to more clearly distinguish this.

Gweaton commented on app/presenters/queries/embedded_content_presenter.rb:
```diff
@@ -0,0 +1,43 @@
+module Presenters
+  module Queries
+    class EmbeddedContentPresenter
+      def self.present(target_edition_id, host_edition)
```
Should the second parameter be `host_editions` (plural) or are we only passing one edition here? In the test we're passing multiple - should this be renamed, or am I misunderstanding here?

tahb replied:
Yes good spot! Thank you.

Gweaton commented on app/queries/get_embedded_content.rb:
```diff
@@ -0,0 +1,39 @@
+      Presenters::Queries::EmbeddedContentPresenter.new(
+        target_content_id,
+        host_editions,
+      ).present
```
Can/should we use the `.present` class method here, or is there a reason we need to use the instance method?

tahb replied:
I think I followed [this Presenter as a template](https://github.com/alphagov/publishing-api/blob/6cee925ae331e7cc059b1432033df8d1d343a103/app/presenters/queries/link_set_presenter.rb#L12) which had the instance method and then I didn't have a test to drive me to add the `self.present(…)` that calls the initialiser internally. I see now that other presenters are doing that so I'll add that too to keep it consistent 👍

Gweaton commented on app/queries/get_embedded_content.rb:
```diff
@@ -0,0 +1,39 @@
+  private
+
+    def host_editions
+      @host_editions ||= Edition.live
```
Would we ever want people to be able to embed content in draft docs e.g. for previewing what's being published? Or are we starting with `live` for now?

tahb replied:
Embedding should work for draft documents so there is an argument that content block publishers might want to see ALL documents regardless of state.

[For our use case I've gone back to clarify what we need as our default case](https://gds.slack.com/archives/C062Z1FA2ES/p1724239210827159). 

In any case I can imagine that consumers might want embedded content in either state. I'll open this up to everything and then we can filter it down later 👍

tahb replied:
I've changed it to do draft and live, but not superseded or unpublished (though we don't have a factory for the latter).

Gweaton commented on spec/queries/get_embedded_content_spec.rb:
```diff
@@ -0,0 +1,130 @@
+      end
+    end
+
+    context "when the target_content_id does not match a live edition" do
```
Is it more accurate to say something like `when the target_content is embedded in non-live editions`...`it does not include the non-live editions in the results`?

I was initially reading this as us searching for the `content_block` via the `target_content_id`, so this might make that a bit clearer.

tahb replied:
Yes this wasn't worded well was it! As i'm changing the behaviour to include draft and published I'll revamp the tests in a new commit which hopefully makes it clearer 👍

Gweaton reviewed (commented):
This looks good to me - just a couple of small questions/comments. Well done for wrangling that complex query!

richardTowers commented on app/queries/get_embedded_content.rb:
```diff
@@ -0,0 +1,39 @@
+        .joins("LEFT JOIN links AS primary_links ON primary_links.edition_id = editions.id AND primary_links.link_type = 'primary_publishing_organisation'")
+        .joins("LEFT JOIN documents ON documents.content_id = primary_links.target_content_id")
+        .joins("LEFT JOIN editions AS org_editions ON org_editions.document_id = documents.id")
+        .where(links: { target_content_id: }) # TODO: Inlcude `link_type: "embedded"`` when it is supported.
```
I believe it is supported, only it's called "embed", rather than "embedded" (added in 372952b1efb68b3d8f78b855d9aa3202a46fd488). Unless I misunderstand?

tahb replied:
Oh super, I didn't realise this had happened already. Thank you 🙌

richardTowers reviewed (commented):
Just one drive by comment - I'll leave it to George to do the proper review. Generally looks good to me. Agree with the logic behind making this a separate endpoint.

richardTowers commented:
Ah, I think I know why [the subquery](https://github.com/alphagov/publishing-api/blob/3f4cc8523b1713aefc1e8f70950e1a81cfa56f22/app/presenters/queries/content_item_presenter.rb#L190) performed so terribly.

[This line](https://github.com/alphagov/publishing-api/blob/3f4cc8523b1713aefc1e8f70950e1a81cfa56f22/app/presenters/queries/content_item_presenter.rb#L195) is ambiguous about what it's referring to as `content_id` - should it be `link_sets.content_id = link_sets.content_id` or `link_sets.content_id = subquery.content_id`. To us it's obvious it should be the latter, but I suspect Postgres is doing the former.

You might find you can get good performance if you update it to `link_sets.content_id = subquery.content_id` to be explicit.

tahb commented:
@Gweaton Thank you for the review! This is definitely in a better place.

I've addressed everything and notably we now:

- return draft _and_ published host content now
- only check for links with the `embed` type and updated/corrected tests

Ready for another review please! 🙏

tahb commented on app/queries/get_embedded_content.rb:
```diff
@@ -0,0 +1,44 @@
+
+    def initialize(target_content_id)
+      self.target_content_id = target_content_id
+      self.states = %i[draft published]
```
I haven't gone as far as making these fully configurable as I can't verify whether we actually need it! That said this is hopefully where it can be injected in from a query param in future 🤞

Gweaton reviewed (approved):
Thanks for making those changes - looks good to me!
